src/vs/workbench/contrib/inlineChat/browser/inlineChatController.ts

```diff
@@ -234,6 +244,54 @@ export class InlineChatController implements IEditorContribution {
 		return INLINE_CHAT_ID;
 	}
 
+	private _installProviderEnablementListeners() {
+		this._store.add(this._inlineChatService.onDidChangeProviders(async (e) => {
+			if (e.added) {
+				this._providerEnablementEvents.set(e.added, e.added.onDidChangeEnablementStatus(async () => {
+					if ((await this._enabledProvider) === e.added) {
+						if (this._session?.provider === e.added) {
+							await this.cancelSession();
+						}
+					}
+					this._enabledProvider = this._refreshAllProviderEnablements(this._editor);
+				}));
+				this._enabledProvider = this._refreshAllProviderEnablements(this._editor);
+			}
+			if (e.removed) {
+				if ((await this._enabledProvider) === e.removed) {
+					this._enabledProvider = Promise.resolve(undefined);
+				}
+				this._providerEnablementEvents.deleteAndDispose(e.removed);
+			}
+		}));
+
+	}
+
+	private async _refreshAllProviderEnablements(editor: ICodeEditor) {
+		let reason: { reason: string } | undefined;
+		for (const provider of this._inlineChatService.getAllProvider()) {
+			const status = await this._getProviderEnablementStatus(editor, provider);
+			if (status === true) {
+				return provider;
+			}
+			reason = status;
+		}
+		return reason;
+	}
+
+	private async _getProviderEnablementStatus(editor: ICodeEditor, provider: IInlineChatSessionProvider) {
+
+		const currentFileURI = editor.getModel()?.uri;
+		if (!currentFileURI) {
+			return { reason: 'No file open' };
+		}
+		const enablementStatus = await provider.provideEnablementStatus(currentFileURI, CancellationToken.None);
+		if (enablementStatus === null || enablementStatus === undefined || enablementStatus === true) {
+			return true;
+		}
+		return enablementStatus;
+	}
```

Comment:
I would suggest to encapsulte the logic of picking a provider into a separate class which is just consumed in this controller.